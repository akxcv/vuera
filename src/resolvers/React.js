/* eslint-disable prefer-object-spread/prefer-object-spread */
import React from 'react'
import { VueWrapper } from '../'
import isReactComponent from '../utils/isReactComponent'

export default function ReactResolver (component) {
  return isReactComponent(component)
    ? component
    : props => <VueWrapper {...props} component={component} />
}

/**
 * This function gets imported by the babel plugin. It wraps a suspected React element and, if it
 * isn't a valid React element, wraps it into a Vue container.
 */
export function babelReactResolver (component, props, children) {
  return isReactComponent(component)
    ? React.createElement(component, props, children)
    : React.createElement(VueWrapper, Object.assign({ component }, props), children)
}
