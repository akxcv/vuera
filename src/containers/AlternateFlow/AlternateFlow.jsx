import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Prepare from '@components/Contents/FaceScan/Prepare';
import { SelfiesTip, Face, TurnHead } from '@components/Contents/AlternateFlow';

export default class AlternateFlow extends Component {
    static propTypes = {
        onNextStep: PropTypes.func,
        onGoBack: PropTypes.func
    };

    static defaultProps = {
        onNextStep: () => null
    };

    constructor(props) {
        super(props);

        this.state = {
            step: 0,
            totalSteps: 4,
            smile: true
        };

        this.element = null;

        this.handleNextStep = this.handleNextStep.bind(this);
        this.handleGoBack = this.handleGoBack.bind(this);
        this.handleCapture = this.handleCapture.bind(this);
    }

    componentDidMount() {
        const random = Math.floor(Math.random() * 2);
        this.setState({ smile: !random });
    }

    /**
     * Go to the next step
     * @return {Void}
     */
    handleNextStep() {
        const { totalSteps, step } = this.state;
        if (step < totalSteps - 1) {
            this.setState(({ step }) => ({ step: step + 1 }));
        } else {
            this.props.onNextStep({ completed: true });
        }
    }

    /**
     * Go back to the previous step
     * @return {Void}
     */
    handleGoBack() {
        const { step } = this.state;
        if (step) {
            this.setState(({ step }) => ({ step: step - 1 }));
        } else {
            this.props.onGoBack();
        }
    }

    /**
     * Handle the capture
     * @return {Void}
     */
    handleCapture() {
        this.element.click();
        this.handleNextStep();
    }

    render() {
        const { step, smile } = this.state;
        return (
            <div>
                {!step && (
                    <Prepare
                        onGoBack={this.handleGoBack}
                        onNextStep={this.handleNextStep}
                    />
                )}
                {step === 1 && (
                    <SelfiesTip
                        onGoBack={this.handleGoBack}
                        onNextStep={this.handleNextStep}
                    />
                )}
                {step === 2 && (
                    <Face smile={smile} onCapture={this.handleCapture} />
                )}
                {step === 3 && <TurnHead onCapture={this.handleCapture} />}
                <input
                    type="file"
                    name="image"
                    accept="image/*"
                    capture="user"
                    ref={ref => (this.element = ref)}
                    style={{ opacity: 0, zIindex: 99 }}
                />
            </div>
        );
    }
}
