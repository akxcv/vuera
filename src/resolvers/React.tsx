/* eslint-disable prefer-object-spread/prefer-object-spread */
import React from "react";
import VueWrapper from "../wrappers/Vue";
import isReactComponent from "../utils/isReactComponent";

export default function ReactResolver(component: any) {
  return isReactComponent(component)
    ? component
    : (props: any) => <VueWrapper {...props} component={component} />;
}

/**
 * This function gets imported by the babel plugin. It wraps a suspected React element and, if it
 * isn't a valid React element, wraps it into a Vue container.
 */
export function babelReactResolver(component: any, props: any, children: any) {
  return isReactComponent(component)
    ? React.createElement(component, props, children)
    : React.createElement(
        VueWrapper,
        Object.assign({ component }, props),
        children
      );
}
