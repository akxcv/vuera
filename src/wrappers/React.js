import React from 'react'
import ReactDOM from 'react-dom'
import VueWrapper from './Vue'

const makeReactContainer = Component => {
  return class ReactInVue extends React.Component {
    static displayName = `ReactInVue${Component.displayName || Component.name || 'Component'}`

    constructor (props) {
      super(props)

      /**
       * We create a stateful component in order to attach a ref on it. We will use that ref to
       * update component's state, which seems better than re-rendering the whole thing with
       * ReactDOM.
       */
      this.state = { ...props }
    }

    wrapVueChildren (children) {
      return {
        render: createElement => createElement('div', children),
      }
    }

    render () {
      const {
        children,
        // Vue attaches an event handler, but it is missing an event name, so
        // it ends up using an empty string. Prevent passing an empty string
        // named prop to React.
        '': _invoker,
        ...rest
      } = this.state
      const wrappedChildren = this.wrapVueChildren(children)

      return (
        <Component {...rest}>
          {children && <VueWrapper component={wrappedChildren} />}
        </Component>
      )
    }
  }
}

export default {
  props: ['component', 'passedProps'],
  render (createElement) {
    return createElement('div')
  },
  methods: {
    mountReactComponent (component) {
      const Component = makeReactContainer(component)
      const children = this.$slots.default !== undefined ? { children: this.$slots.default } : {}
      ReactDOM.render(
        <Component
          {...this.$props.passedProps}
          {...this.$attrs}
          {...this.$listeners}
          {...children}
          ref={ref => (this.reactComponentRef = ref)}
        />,
        this.$el
      );
    },
  },
  mounted () {
    this.mountReactComponent(this.$props.component)
  },
  beforeDestroy () {
    ReactDOM.unmountComponentAtNode(this.$el)
  },
  updated () {
    /**
     * AFAIK, this is the only way to update children. It doesn't seem to be possible to watch
     * `$slots` or `$children`.
     */
    if (this.$slots.default !== undefined) {
      this.reactComponentRef.setState({ children: this.$slots.default })
    } else {
      this.reactComponentRef.setState({ children: null })
    }
  },
  inheritAttrs: false,
  watch: {
    $attrs: {
      handler () {
        this.reactComponentRef.setState({ ...this.$attrs })
      },
      deep: true,
    },
    '$props.component': {
      handler (newValue) {
        this.mountReactComponent(newValue)
      },
    },
    $listeners: {
      handler () {
        this.reactComponentRef.setState({ ...this.$listeners })
      },
      deep: true,
    },
    '$props.passedProps': {
      handler () {
        this.reactComponentRef.setState({ ...this.$props.passedProps })
      },
      deep: true,
    },
  },
}
