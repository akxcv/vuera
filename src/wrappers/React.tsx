import React from "react";
import ReactDOM from "react-dom";
import VueWrapper from "./Vue";

const makeReactContainer = (Component: any) =>
  class ReactInVue extends React.Component {
    static displayName = `ReactInVue${
      Component.displayName || Component.name || "Component"
    }`;

    constructor(props: any) {
      super(props);

      /**
       * We create a stateful component in order to attach a ref on it. We will use that ref to
       * update component's state, which seems better than re-rendering the whole thing with
       * ReactDOM.
       */
      this.state = { ...props };
    }

    wrapVueChildren(children: any) {
      // console.log("wrapVueChildren: ", children);
      if (children)
        return {
          render: (createElement: any) => createElement("div", children),
        };
      return null;
    }

    render() {
      const {
        children,
        // Vue attaches an event handler, but it is missing an event name, so
        // it ends up using an empty string. Prevent passing an empty string
        // named prop to React.
        "": _invoker,
        ...rest
      } = (this as any).state;
      const wrappedChildren = this.wrapVueChildren(children);

      const VueWrapperRender = VueWrapper as unknown as (props: {
        component: any;
      }) => JSX.Element;

      if ("ReactInVueTestAA" === ReactInVue.displayName) {
        // console.log("THIS IS IT!", Component.render);
      }

      // console.log("wrappedChildren: ", wrappedChildren);

      return (
        <Component {...rest}>
          {wrappedChildren && <VueWrapperRender component={wrappedChildren} />}
        </Component>
      );
    }
  };

export default {
  name: "ReactInVueRawVueComp",
  props: ["component", "passedProps"],
  render(createElement: any) {
    (this as any).createElement = createElement; // save for later
    return createElement("div", { ref: "react" });
  },
  methods: {
    mountReactComponent(comp: any) {
      const s = this as any;
      // console.log("before creating NewComp");
      // console.log("Name: ", comp.name || comp.displayName);
      // console.log(comp);

      const children =
        s.$slots.default !== undefined ? { children: s.$slots.default } : {};

      let NewComp: React.FC;
      // if (!comp.functional) {
        const Component = makeReactContainer(comp);
        NewComp = (props: any) => (
          <Component
            {...props}
            ref={(ref: any) => (s.reactComponentRef = ref)}
          />
        );
      // } else {
      //   NewComp = (props: any) => <>{comp.render(s.createElement, props)}</>;
      // }
      // NewComp = (props: any) => <>{comp.render(s.createElement, props)}</>;

      ReactDOM.render(
        <NewComp
          {...s.$props.passedProps}
          {...s.$attrs}
          {...s.$listeners}
          {...children}
        />,
        s.$refs.react
      );
      // console.log("after creating NewComp");
    },
  },
  mounted() {
    (this as any).mountReactComponent((this as any).$props.component);
  },
  beforeDestroy() {
    ReactDOM.unmountComponentAtNode((this as any).$refs.react);
  },
  updated() {
    /**
     * AFAIK, this is the only way to update children. It doesn't seem to be possible to watch
     * `$slots` or `$children`.
     */
    if ((this as any).$slots.default !== undefined) {
      (this as any).reactComponentRef.setState({
        children: (this as any).$slots.default,
      });
    } else {
      (this as any).reactComponentRef.setState({ children: null });
    }
  },
  inheritAttrs: false,
  watch: {
    $attrs: {
      handler() {
        (this as any).reactComponentRef.setState({ ...(this as any).$attrs });
      },
      deep: true,
    },
    "$props.component": {
      handler(newValue: any) {
        (this as any).mountReactComponent(newValue);
      },
    },
    $listeners: {
      handler() {
        (this as any).reactComponentRef.setState({
          ...(this as any).$listeners,
        });
      },
      deep: true,
    },
    "$props.passedProps": {
      handler() {
        (this as any).reactComponentRef.setState({
          ...(this as any).$props.passedProps,
        });
      },
      deep: true,
    },
  },
};
