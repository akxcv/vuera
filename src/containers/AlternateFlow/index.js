import AlternateFlow from './AlternateFlow';
export default AlternateFlow;
		