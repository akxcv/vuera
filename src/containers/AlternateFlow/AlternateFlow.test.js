import React from 'react';
import renderer from 'react-test-renderer';
import AlternateFlow from './AlternateFlow';

describe('AlternateFlow', () => {
    it('renders', () => {
        const tree = renderer.create(<AlternateFlow />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
