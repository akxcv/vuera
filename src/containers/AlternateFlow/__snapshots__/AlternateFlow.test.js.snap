// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlternateFlow renders 1`] = `
<div>
  <div
    className=""
  >
    <header
      className={undefined}
      style={
        Object {
          "position": "fixed",
        }
      }
    >
      <div
        className="b-container"
      >
        <button
          className={undefined}
          onClick={[Function]}
          type="button"
        />
        <h3
          className={undefined}
        >
          Face verification
        </h3>
      </div>
    </header>
    <div
      className={undefined}
    >
      <div
        className={undefined}
      >
        <div
          className="b-title"
        >
          We are now going to perform a face verification.
        </div>
        <div
          className={undefined}
        >
          <div>
            Please don't have anything covering your face.
          </div>
          <img
            alt=""
            src="images/tips/facescan/no-hat-classes.png"
          />
          <span>
            Sunglasses and hats for example.
          </span>
        </div>
      </div>
    </div>
    <div
      className={undefined}
    >
      <div
        className={undefined}
        style={
          Object {
            "padding": "20px 0",
          }
        }
      >
        <div
          className={undefined}
        >
          <button
            className=""
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Next
          </button>
        </div>
      </div>
    </div>
  </div>
  <input
    accept="image/*"
    capture="user"
    name="image"
    style={
      Object {
        "opacity": 0,
        "zIindex": 99,
      }
    }
    type="file"
  />
</div>
`;
